File: solutions/python/matrix/1/matrix.py
class Matrix:
    data: list[list[int]]

    def __init__(self, matrix_string: str):
        self.data = [
            [int(cell) for cell in row.split()]
            for row in matrix_string.split("\n")
        ]

    def row(self, index: int) -> list[int]:
        return self.data[index - 1]

    def column(self, index: int) -> list[int]:
        return [row[index - 1] for row in self.data]

File: solutions/haskell/saddle-points/2/src/Matrix.hs
{-# LANGUAGE BlockArguments #-}

module Matrix (saddlePoints) where

import Data.Array

saddlePoints :: (Ord e) => Array (Int, Int) e -> [(Int, Int)]
saddlePoints matrix =
    [(i, j) | i <- i_range, j <- find_max_indexes i, is_min_in_column i j]
  where
    ((i_min, j_min), (i_max, j_max)) = bounds matrix
    i_range = [i_min .. i_max]
    row i = ixmap (j_min, j_max) (i,) matrix
    column j = ixmap (i_min, i_max) (,j) matrix

    find_max_indexes i = [j | (j, v) <- assocs ri, v == maxv]
      where
        ri = row i
        maxv = maximum ri

    is_min_in_column i j = all (>= v) $ column j where v = matrix ! (i, j)
